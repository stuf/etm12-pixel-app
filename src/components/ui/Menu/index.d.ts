/**
 * Represents a single menu item, that itself may contain
 * menu items
 */
export interface MenuItem {
  label: string;
  accelerator?: any;
  items?: MenuItem[];
}

export interface Props {
  items: MenuItem[];
}
