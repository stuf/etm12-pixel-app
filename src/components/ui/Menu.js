import * as React from 'karet';
import * as U from 'karet.util';
import * as T from './Menu';

import styles from './Menu.module.scss';

/**
 * @param {T.Props} props
 */
function Menu({ items }) {
  return (
    <section className={styles.root}>
      <ul className={styles.itemList}>
        {U.thru(
          items,
          U.mapElems((it, i) => (
            <li
              {...{
                key: `menuitem-${i}`,
                className: styles.item,
              }}
            >
              {U.view('label', it)}
            </li>
          )),
        )}
      </ul>
    </section>
  );
}

export default Menu;

//
