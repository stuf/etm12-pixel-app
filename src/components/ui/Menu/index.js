/* eslint no-unused-vars: [1, {"varsIgnorePattern": "[K|T|Model]"}] */
import * as React from 'karet';
import * as U from 'karet.util';

import * as T from './index.d';
import styles from './index.module.scss';

/**
 * @param {T.Props} props
 * @return {T.Component}
 */
function Menu({ items }) {
  return (
    <section className={styles.root}>
      <ul className={styles.itemList}>
        {U.thru(
          items,
          U.mapElems((it, i) => (
            <li
              {...{
                key: `menuitem-${i}`,
                className: U.cns(
                  styles.item,
                  U.when(U.view('disabled', it), styles.disabled),
                ),
              }}
            >
              {U.view('label', it)}
            </li>
          )),
        )}
      </ul>
    </section>
  );
}

export default Menu;

//
