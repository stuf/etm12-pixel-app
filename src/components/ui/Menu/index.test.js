import * as React from 'karet';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Menu from './index';

it('renders without crashing', () => {
  const w1 = shallow(<Menu items={[]} />);

  const w2 = shallow(
    <Menu
      items={[{ label: 'test 1' }, { label: 'test 2' }, { label: 'test 3' }]}
    />,
  );

  expect(toJson(w1)).toMatchSnapshot('menu with no items');
  expect(toJson(w2)).toMatchSnapshot('menu with items');
});
