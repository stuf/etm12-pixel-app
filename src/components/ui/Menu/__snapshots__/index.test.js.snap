// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing: menu with items 1`] = `
<section
  className="root"
>
  <ul
    className="itemList"
  >
    <li
      className="item"
      key="menuitem-0"
    >
      test 1
    </li>
    <li
      className="item"
      key="menuitem-1"
    >
      test 2
    </li>
    <li
      className="item"
      key="menuitem-2"
    >
      test 3
    </li>
  </ul>
</section>
`;

exports[`renders without crashing: menu with no items 1`] = `
<section
  className="root"
>
  <ul
    className="itemList"
  />
</section>
`;
